### PR TITLE
[Hotfix][Core][Flink] SeaTunnel flink engine support application mode on yarn

### PR DESCRIPTION
**Pull request by TyrantLucifer**
### Purpose of this pull request

SeaTunnel flink engine support application mode on yarn

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Verify whether the Flink on Yarn application mode can be submitted normally

- flink 1.13

```shell
bin/start-seatunnel-flink-13-connector-v2.sh --config config/v2.streaming.conf.template --target yarn-application -e run-application --name st-test
```

image

- flink 1.16

```shell
bin/start-seatunnel-flink-15-connector-v2.sh --config config/v2.streaming.conf.template --target yarn-application -e run-application --name st-test
```

image

### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  2. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
  3. Add ci label in [label-scope-conf](https://github.com/apache/seatunnel/blob/dev/.github/workflows/labeler/label-scope-conf.yml)
  4. Add e2e testcase in [seatunnel-e2e](https://github.com/apache/seatunnel/tree/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/)
  5. Update connector [plugin_config](https://github.com/apache/seatunnel/blob/dev/config/plugin_config)
* [ ] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

### CODE REVIEW

**Comment by TyrantLucifer** [comment]
@hailin0 @Hisoka-X @Carl-Zhou-CN PTAL

**Comment by hailin0** [comment]
Please try again ci

**Comment by Adamyuanyuan** [comment]
root@hadoop-master1:/mnt/apache-seatunnel-2.3.8-SNAPSHOT# ./bin/start-seatunnel-flink-15-connector-v2.sh --config ./config/mysql/mysql2hive2_partitioned.conf --target yarn-application -e run-application --name st-test                                                                                               
Execute SeaTunnel Flink Job: ${FLINK_HOME}/bin/flink run-application --target yarn-application -Dyarn.ship-files="./config/mysql/mysql2hive2_partitioned.conf" -Dyarn.ship-archives=runtime.tar.gz -Dyarn.application.name=st-test -c org.apache.seatunnel.core.starter.flink.SeaTunnelFlink /mnt/apache-seatunnel-2.3.8-SNAPSHOT/starter/seatunnel-flink-15-starter.jar --config ./config/mysql/mysql2hive2_partitioned.conf --name st-test --deploy-mode run-application          
2024-10-09 10:58:12,913 INFO  org.apache.flink.yarn.cli.FlinkYarnSessionCli                [] - Found Yarn properties file under /tmp/.yarn-properties-root.
2024-10-09 10:58:12,913 INFO  org.apache.flink.yarn.cli.FlinkYarnSessionCli                [] - Found Yarn properties file under /tmp/.yarn-properties-root.
2024-10-09 10:58:13,067 INFO  org.apache.hadoop.yarn.client.DefaultNoHARMFailoverProxyProvider [] - Connecting to ResourceManager at hadoop-master1.orb.local/172.31.0.5:8032                                                                                                                                           
2024-10-09 10:58:13,170 INFO  org.apache.flink.yarn.YarnClusterDescriptor                  [] - No path for the flink jar passed. Using the location of class org.apache.flink.yarn.YarnClusterDescriptor to locate the jar                                                                                             
                                                                                                                                                            
------------------------------------------------------------                                                                                                
 The program finished with the following exception:                                                                                                         
                                                                                                                                                            
java.lang.IllegalArgumentException: Directories or non-archive files are included.                                                                          
        at org.apache.flink.util.Preconditions.checkArgument(Preconditions.java:138)                                                                        
        at org.apache.flink.yarn.YarnClusterDescriptor.addShipArchives(YarnClusterDescriptor.java:306)                                                      
        at java.util.Optional.ifPresent(Optional.java:159)                                                                                                  
        at org.apache.flink.yarn.YarnClusterDescriptor.<init>(YarnClusterDescriptor.java:197)                                                               
        at org.apache.flink.yarn.YarnClusterClientFactory.getClusterDescriptor(YarnClusterClientFactory.java:88)                                            
        at org.apache.flink.yarn.YarnClusterClientFactory.createClusterDescriptor(YarnClusterClientFactory.java:60)                                         
        at org.apache.flink.yarn.YarnClusterClientFactory.createClusterDescriptor(YarnClusterClientFactory.java:42)                                         
        at org.apache.flink.client.deployment.application.cli.ApplicationClusterDeployer.run(ApplicationClusterDeployer.java:63)                            
        at org.apache.flink.client.cli.CliFrontend.runApplication(CliFrontend.java:212)                                                                     
        at org.apache.flink.client.cli.CliFrontend.parseAndRun(CliFrontend.java:1098)                                                                       
        at org.apache.flink.client.cli.CliFrontend.lambda$mainInternal$9(CliFrontend.java:1189)                                                             
        at java.security.AccessController.doPrivileged(Native Method)                                                                                       
        at javax.security.auth.Subject.doAs(Subject.java:422)                                                                                               
        at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1899)                                                             
        at org.apache.flink.runtime.security.contexts.HadoopSecurityContext.runSecured(HadoopSecurityContext.java:41)                                       
        at org.apache.flink.client.cli.CliFrontend.mainInternal(CliFrontend.java:1189)                                                                      
        at org.apache.flink.client.cli.CliFrontend.main(CliFrontend.java:1157)

**Comment by TyrantLucifer** [comment]
> runtime.tar.gz

Check if runtime.tar.gz is generated normally in the ST directory

**Comment by TyrantLucifer** [comment]
> root@hadoop-master1:/mnt/apache-seatunnel-2.3.8-SNAPSHOT# ./bin/start-seatunnel-flink-15-connector-v2.sh --config ./config/mysql/mysql2hive2_partitioned.conf --target yarn-application -e run-application --name st-test Execute SeaTunnel Flink Job: ${FLINK_HOME}/bin/flink run-application --target yarn-application -Dyarn.ship-files="./config/mysql/mysql2hive2_partitioned.conf" -Dyarn.ship-archives=runtime.tar.gz -Dyarn.application.name=st-test -c org.apache.seatunnel.core.starter.flink.SeaTunnelFlink /mnt/apache-seatunnel-2.3.8-SNAPSHOT/starter/seatunnel-flink-15-starter.jar --config ./config/mysql/mysql2hive2_partitioned.conf --name st-test --deploy-mode run-application 2024-10-09 10:58:12,913 INFO org.apache.flink.yarn.cli.FlinkYarnSessionCli [] - Found Yarn properties file under /tmp/.yarn-properties-root. 2024-10-09 10:58:12,913 INFO org.apache.flink.yarn.cli.FlinkYarnSessionCli [] - Found Yarn properties file under /tmp/.yarn-properties-root. 2024-10-09 10:58:13,067 INFO org.apache.hadoop.yarn.client.DefaultNoHARMFailoverProxyProvider [] - Connecting to ResourceManager at hadoop-master1.orb.local/172.31.0.5:8032 2024-10-09 10:58:13,170 INFO org.apache.flink.yarn.YarnClusterDescriptor [] - No path for the flink jar passed. Using the location of class org.apache.flink.yarn.YarnClusterDescriptor to locate the jar
> 
> The program finished with the following exception:
> 
> java.lang.IllegalArgumentException: Directories or non-archive files are included. at org.apache.flink.util.Preconditions.checkArgument(Preconditions.java:138) at org.apache.flink.yarn.YarnClusterDescriptor.addShipArchives(YarnClusterDescriptor.java:306) at java.util.Optional.ifPresent(Optional.java:159) at org.apache.flink.yarn.YarnClusterDescriptor.(YarnClusterDescriptor.java:197) at org.apache.flink.yarn.YarnClusterClientFactory.getClusterDescriptor(YarnClusterClientFactory.java:88) at org.apache.flink.yarn.YarnClusterClientFactory.createClusterDescriptor(YarnClusterClientFactory.java:60) at org.apache.flink.yarn.YarnClusterClientFactory.createClusterDescriptor(YarnClusterClientFactory.java:42) at org.apache.flink.client.deployment.application.cli.ApplicationClusterDeployer.run(ApplicationClusterDeployer.java:63) at org.apache.flink.client.cli.CliFrontend.runApplication(CliFrontend.java:212) at org.apache.flink.client.cli.CliFrontend.parseAndRun(CliFrontend.java:1098) at org.apache.flink.client.cli.CliFrontend.lambda$mainInternal$9(CliFrontend.java:1189) at java.security.AccessController.doPrivileged(Native Method) at javax.security.auth.Subject.doAs(Subject.java:422) at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1899) at org.apache.flink.runtime.security.contexts.HadoopSecurityContext.runSecured(HadoopSecurityContext.java:41) at org.apache.flink.client.cli.CliFrontend.mainInternal(CliFrontend.java:1189) at org.apache.flink.client.cli.CliFrontend.main(CliFrontend.java:1157)

```bash
if [ ! -f "${APP_DIR}/runtime.tar.gz" ];then

  directories=("connectors" "lib" "plugins")

  existing_dirs=()

  for dir in "${directories[@]}"; do
      if [ -d "$dir" ]; then
          existing_dirs+=("$dir")
      fi
  done

  if [ ${#existing_dirs[@]} -eq 0 ]; then
      echo "[connectors,lib,plugins] not existed, skip generate runtime.tar.gz"
  else
      tar -zcvf runtime.tar.gz "${existing_dirs[@]}"
  fi
fi
```

**Comment by Adamyuanyuan** [comment]
1. 我手动检查路径，发现根路径没有runtime.tar.gz；
2. 我手动打包，就可以提交了； tar -zcvf runtime.tar.gz connectors lib plugins
3. 提交上去，开始报这个错误，跟之前没打这个patch前报错一样

2024-10-10 10:27:26,914 ERROR org.apache.seatunnel.core.starter.SeaTunnel                  [] - Fatal Error, 

2024-10-10 10:27:26,914 ERROR org.apache.seatunnel.core.starter.SeaTunnel                  [] - Please submit bug report in https://github.com/apache/seatunnel/issues

2024-10-10 10:27:26,914 ERROR org.apache.seatunnel.core.starter.SeaTunnel                  [] - Reason:ErrorCode:[COMMON-22], ErrorDescription:[SeaTunnel read file 'config/mysql/mysql2hive2_partitioned.conf' failed, because it not existed.] 

2024-10-10 10:27:26,914 ERROR org.apache.seatunnel.core.starter.SeaTunnel                  [] - Exception StackTrace:org.apache.seatunnel.common.exception.SeaTunnelRuntimeException: ErrorCode:[COMMON-22], ErrorDescription:[SeaTunnel read file 'config/mysql/mysql2hive2_partitioned.conf' failed, because it not existed.]
	at org.apache.seatunnel.common.exception.CommonError.fileNotExistFailed(CommonError.java:86)
	at org.apache.seatunnel.core.starter.utils.FileUtils.checkConfigExist(FileUtils.java:66)
	at org.apache.seatunnel.core.starter.flink.command.FlinkTaskExecuteCommand.execute(FlinkTaskExecuteCommand.java:50)
	at org.apache.seatunnel.core.starter.SeaTunnel.run(SeaTunnel.java:40)
	at org.apache.seatunnel.core.starter.flink.SeaTunnelFlink.main(SeaTunnelFlink.java:34)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.flink.client.program.PackagedProgram.callMainMethod(PackagedProgram.java:355)
	at org.apache.flink.client.program.PackagedProgram.invokeInteractiveModeForExecution(PackagedProgram.java:222)
	at org.apache.flink.client.ClientUtils.executeProgram(ClientUtils.java:105)
	at org.apache.flink.client.deployment.application.ApplicationDispatcherBootstrap.runApplicationEntryPoint(ApplicationDispatcherBootstrap.java:301)
	at org.apache.flink.client.deployment.application.ApplicationDispatcherBootstrap.lambda$runApplicationAsync$2(ApplicationDispatcherBootstrap.java:254)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at org.apache.flink.runtime.concurrent.pekko.ActorSystemScheduledExecutorAdapter$ScheduledFutureTask.run(ActorSystemScheduledExecutorAdapter.java:172)
	at org.apache.flink.runtime.concurrent.ClassLoadingUtils.runWithContextClassLoader(ClassLoadingUtils.java:68)
	at org.apache.flink.runtime.concurrent.ClassLoadingUtils.lambda$withContextClassLoader$0(ClassLoadingUtils.java:41)
	at org.apache.pekko.dispatch.TaskInvocation.run(AbstractDispatcher.scala:59)
	at org.apache.pekko.dispatch.ForkJoinExecutorConfigurator$PekkoForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:57)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:175)
 
2024-10-10 10:27:26,914 ERROR org.apache.seatunnel.core.starter.SeaTunnel                  [] - 


4. 我再次打包 tar -zcvf runtime.tar.gz connectors lib plugins config ，加入了 config，发现还是一样找不到 'config/mysql/mysql2hive2_partitioned.conf'，奇怪了。

**Comment by TyrantLucifer** [comment]
> 1. 我手动检查路径，发现根路径没有runtime.tar.gz；
> 2. 我手动打包，就可以提交了； tar -zcvf runtime.tar.gz connectors lib plugins
> 3. 提交上去，开始报这个错误，跟之前没打这个patch前报错一样
> 
> 2024-10-10 10:27:26,914 ERROR org.apache.seatunnel.core.starter.SeaTunnel [] - Fatal Error,
> 
> 2024-10-10 10:27:26,914 ERROR org.apache.seatunnel.core.starter.SeaTunnel [] - Please submit bug report in https://github.com/apache/seatunnel/issues
> 
> 2024-10-10 10:27:26,914 ERROR org.apache.seatunnel.core.starter.SeaTunnel [] - Reason:ErrorCode:[COMMON-22], ErrorDescription:[SeaTunnel read file 'config/mysql/mysql2hive2_partitioned.conf' failed, because it not existed.]
> 
> 2024-10-10 10:27:26,914 ERROR org.apache.seatunnel.core.starter.SeaTunnel [] - Exception StackTrace:org.apache.seatunnel.common.exception.SeaTunnelRuntimeException: ErrorCode:[COMMON-22], ErrorDescription:[SeaTunnel read file 'config/mysql/mysql2hive2_partitioned.conf' failed, because it not existed.] at org.apache.seatunnel.common.exception.CommonError.fileNotExistFailed(CommonError.java:86) at org.apache.seatunnel.core.starter.utils.FileUtils.checkConfigExist(FileUtils.java:66) at org.apache.seatunnel.core.starter.flink.command.FlinkTaskExecuteCommand.execute(FlinkTaskExecuteCommand.java:50) at org.apache.seatunnel.core.starter.SeaTunnel.run(SeaTunnel.java:40) at org.apache.seatunnel.core.starter.flink.SeaTunnelFlink.main(SeaTunnelFlink.java:34) at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) at java.lang.reflect.Method.invoke(Method.java:498) at org.apache.flink.client.program.PackagedProgram.callMainMethod(PackagedProgram.java:355) at org.apache.flink.client.program.PackagedProgram.invokeInteractiveModeForExecution(PackagedProgram.java:222) at org.apache.flink.client.ClientUtils.executeProgram(ClientUtils.java:105) at org.apache.flink.client.deployment.application.ApplicationDispatcherBootstrap.runApplicationEntryPoint(ApplicationDispatcherBootstrap.java:301) at org.apache.flink.client.deployment.application.ApplicationDispatcherBootstrap.lambda$runApplicationAsync$2(ApplicationDispatcherBootstrap.java:254) at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) at java.util.concurrent.FutureTask.run(FutureTask.java:266) at org.apache.flink.runtime.concurrent.pekko.ActorSystemScheduledExecutorAdapter$ScheduledFutureTask.run(ActorSystemScheduledExecutorAdapter.java:172) at org.apache.flink.runtime.concurrent.ClassLoadingUtils.runWithContextClassLoader(ClassLoadingUtils.java:68) at org.apache.flink.runtime.concurrent.ClassLoadingUtils.lambda$withContextClassLoader$0(ClassLoadingUtils.java:41) at org.apache.pekko.dispatch.TaskInvocation.run(AbstractDispatcher.scala:59) at org.apache.pekko.dispatch.ForkJoinExecutorConfigurator$PekkoForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:57) at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289) at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056) at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692) at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:175)
> 
> 2024-10-10 10:27:26,914 ERROR org.apache.seatunnel.core.starter.SeaTunnel [] -
> 
> 4. 我再次打包 tar -zcvf runtime.tar.gz connectors lib plugins config ，加入了 config，发现还是一样找不到 'config/mysql/mysql2hive2_partitioned.conf'，奇怪了。

Did you use my branch again to generate full volume packages?

**Comment by TyrantLucifer** [comment]
> > 1. 我手动检查路径，发现根路径没有runtime.tar.gz；
> > 2. 我手动打包，就可以提交了； tar -zcvf runtime.tar.gz connectors lib plugins
> > 3. 提交上去，开始报这个错误，跟之前没打这个patch前报错一样
> > 
> > 2024-10-10 10:27:26,914 ERROR org.apache.seatunnel.core.starter.SeaTunnel [] - Fatal Error,
> > 2024-10-10 10:27:26,914 ERROR org.apache.seatunnel.core.starter.SeaTunnel [] - Please submit bug report in https://github.com/apache/seatunnel/issues
> > 2024-10-10 10:27:26,914 ERROR org.apache.seatunnel.core.starter.SeaTunnel [] - Reason:ErrorCode:[COMMON-22], ErrorDescription:[SeaTunnel read file 'config/mysql/mysql2hive2_partitioned.conf' failed, because it not existed.]
> > 2024-10-10 10:27:26,914 ERROR org.apache.seatunnel.core.starter.SeaTunnel [] - Exception StackTrace:org.apache.seatunnel.common.exception.SeaTunnelRuntimeException: ErrorCode:[COMMON-22], ErrorDescription:[SeaTunnel read file 'config/mysql/mysql2hive2_partitioned.conf' failed, because it not existed.] at org.apache.seatunnel.common.exception.CommonError.fileNotExistFailed(CommonError.java:86) at org.apache.seatunnel.core.starter.utils.FileUtils.checkConfigExist(FileUtils.java:66) at org.apache.seatunnel.core.starter.flink.command.FlinkTaskExecuteCommand.execute(FlinkTaskExecuteCommand.java:50) at org.apache.seatunnel.core.starter.SeaTunnel.run(SeaTunnel.java:40) at org.apache.seatunnel.core.starter.flink.SeaTunnelFlink.main(SeaTunnelFlink.java:34) at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) at java.lang.reflect.Method.invoke(Method.java:498) at org.apache.flink.client.program.PackagedProgram.callMainMethod(PackagedProgram.java:355) at org.apache.flink.client.program.PackagedProgram.invokeInteractiveModeForExecution(PackagedProgram.java:222) at org.apache.flink.client.ClientUtils.executeProgram(ClientUtils.java:105) at org.apache.flink.client.deployment.application.ApplicationDispatcherBootstrap.runApplicationEntryPoint(ApplicationDispatcherBootstrap.java:301) at org.apache.flink.client.deployment.application.ApplicationDispatcherBootstrap.lambda$runApplicationAsync$2(ApplicationDispatcherBootstrap.java:254) at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) at java.util.concurrent.FutureTask.run(FutureTask.java:266) at org.apache.flink.runtime.concurrent.pekko.ActorSystemScheduledExecutorAdapter$ScheduledFutureTask.run(ActorSystemScheduledExecutorAdapter.java:172) at org.apache.flink.runtime.concurrent.ClassLoadingUtils.runWithContextClassLoader(ClassLoadingUtils.java:68) at org.apache.flink.runtime.concurrent.ClassLoadingUtils.lambda$withContextClassLoader$0(ClassLoadingUtils.java:41) at org.apache.pekko.dispatch.TaskInvocation.run(AbstractDispatcher.scala:59) at org.apache.pekko.dispatch.ForkJoinExecutorConfigurator$PekkoForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:57) at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289) at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056) at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692) at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:175)
> > 2024-10-10 10:27:26,914 ERROR org.apache.seatunnel.core.starter.SeaTunnel [] -
> > 
> > 4. 我再次打包 tar -zcvf runtime.tar.gz connectors lib plugins config ，加入了 config，发现还是一样找不到 'config/mysql/mysql2hive2_partitioned.conf'，奇怪了。
> 
> Did you use my branch again to generate full volume packages?

image

image

**Comment by Adamyuanyuan** [comment]
Thanks for the reminder.
I just packaged and replaced the seatunnel-flink-15-starter.jar file. I applied this patch to my 2.3.7 version.
I noticed that there is a related change in the common package as well, so I will try to package everything completely later.

**Comment by TyrantLucifer** [comment]
> Thanks for the reminder. I just packaged and replaced the seatunnel-flink-15-starter.jar file. I applied this patch to my 2.3.7 version. I noticed that there is a related change in the common package as well, so I will try to package everything completely later.

waiting for your reply！

**Comment by hailin0** [comment]
> Thanks for the reminder. I just packaged and replaced the seatunnel-flink-15-starter.jar file. I applied this patch to my 2.3.7 version. I noticed that there is a related change in the common package as well, so I will try to package everything completely later.

Looking forward to your reply

**Review by hailin0 — APPROVED** [review]
LGTM

**Review by Hisoka-X — COMMENTED** [review]
Could you update the doc to share the way of how to submit job to yarn?

**Comment by TyrantLucifer** [comment]
> Could you update the doc to share the way of how to submit job to yarn?

Flink's document submission process is not working well. I plan to rewrite it and complete it in the next PR. Do you think it's okay？

**Comment by Hisoka-X** [comment]
> > Could you update the doc to share the way of how to submit job to yarn?
> 
> Flink's document submission process is not working well. I plan to rewrite it and complete it in the next PR. Do you think it's okay？

Sure, we should complete it before next release. Add a test case for yarn would be better.

File: seatunnel-core/seatunnel-flink-starter/seatunnel-flink-13-starter/src/main/bin/start-seatunnel-flink-13-connector-v2.sh [hunk 1]
```diff
@@ -43,6 +43,25 @@ if [ -f "${CONF_DIR}/seatunnel-env.sh" ]; then
     . "${CONF_DIR}/seatunnel-env.sh"
 fi
 
+if [ ! -f "${APP_DIR}/runtime.tar.gz" ];then
```
Comment (Hisoka-X) [inline seatunnel-core/seatunnel-flink-starter/seatunnel-flink-13-starter/src/main/bin/start-seatunnel-flink-13-connector-v2.sh:46]: If not only one job are submitted at same time, will the files between different tasks conflict?

Comment (TyrantLucifer) [reply seatunnel-core/seatunnel-flink-starter/seatunnel-flink-13-starter/src/main/bin/start-seatunnel-flink-13-connector-v2.sh:46]: Yes, you are right. If user use application mode, runtime.tar.gz should be prepared by user first, this step we will show it in new version docs.

**Comment by TyrantLucifer** [comment]
> > > Could you update the doc to share the way of how to submit job to yarn?
> > 
> > 
> > Flink's document submission process is not working well. I plan to rewrite it and complete it in the next PR. Do you think it's okay？
> 
> Sure, we should complete it before next release. Add a test case for yarn would be better.

#7853

**Review by Hisoka-X — APPROVED** [review]

**Comment by Adamyuanyuan** [comment]
> > Thanks for the reminder. I just packaged and replaced the seatunnel-flink-15-starter.jar file. I applied this patch to my 2.3.7 version. I noticed that there is a related change in the common package as well, so I will try to package everything completely later.
> 
> Looking forward to your reply

In my environment, the dev branch + patch is now passing the tests. Thanks to the support from @TyrantLucifer .

**Pull request merged**